File: UserServer/src/main/java/cc/nobrain/dev/userserver/domain/notice/repository/NoticeRepositoryImpl.java
package cc.nobrain.dev.userserver.domain.notice.repository;

import cc.nobrain.dev.userserver.domain.notice.entity.Notice;
import cc.nobrain.dev.userserver.domain.notice.entity.QNotice;
import com.querydsl.core.types.OrderSpecifier;
import com.querydsl.core.types.Predicate;
import com.querydsl.core.types.Projections;
import com.querydsl.jpa.impl.JPAQueryFactory;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.support.PageableExecutionUtils;

import java.util.List;
import java.util.function.LongSupplier;

@RequiredArgsConstructor
public class NoticeRepositoryImpl implements NoticeRepositoryCustom {

    private final JPAQueryFactory factory;

    @Override
    public List<Notice> findAll(Predicate predicate, OrderSpecifier<?>... orders) {
        return factory.selectFrom(QNotice.notice)
                .where(predicate)
                .orderBy(orders)
                .fetch();
    }

    @Override
    public Page<Notice> findAll(Predicate predicate, Pageable pageable, OrderSpecifier<?>... orders) {
        List query = factory.selectFrom(QNotice.notice)
                .where(predicate)
                .orderBy(orders)
                .offset(pageable.getOffset())
                .limit(pageable.getPageSize())
                .fetch();

        LongSupplier total = () -> factory
                .selectFrom(QNotice.notice)
                .where(predicate)
                .fetchCount();

        return PageableExecutionUtils.getPage(query, pageable, total);
    }

    @Override
    public <T> List<T> findAll(Predicate predicate, Class<T> clazz, OrderSpecifier<?>... orders) {
        return factory.select(Projections.constructor(clazz))
                .from(QNotice.notice)
                .where(predicate)
                .orderBy(orders)
                .fetch();
    }
}

File: UserServer/src/main/java/cc/nobrain/dev/userserver/domain/notice/repository/NoticeRepositoryImpl.kt
package cc.nobrain.dev.userserver.domain.notice.repository

import cc.nobrain.dev.userserver.domain.notice.entity.Notice
import cc.nobrain.dev.userserver.domain.notice.entity.QNotice
import com.querydsl.core.types.OrderSpecifier
import com.querydsl.core.types.Predicate
import com.querydsl.core.types.Projections
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.support.PageableExecutionUtils

class NoticeRepositoryImpl(private val factory: JPAQueryFactory): NoticeRepositoryCustom {

    override fun findAll(predicate: Predicate, vararg orders: OrderSpecifier<*>): List<Notice> =
        factory.selectFrom(QNotice.notice)
            .where(predicate)
            .orderBy(*orders)
            .fetch()

    override fun findAll(predicate: Predicate, pageable: Pageable, vararg orders: OrderSpecifier<*>): Page<Notice> {
        val query = factory.selectFrom(QNotice.notice)
            .where(predicate)
            .orderBy(*orders)
            .offset(pageable.offset)
            .limit(pageable.pageSize.toLong())
            .fetch()

        val total: () -> Long = {
            factory
                .selectFrom(QNotice.notice)
                .where(predicate)
                .fetchCount()
        }

        return PageableExecutionUtils.getPage(query, pageable, total)
    }

    override fun <T> findAll(predicate: Predicate, clazz: Class<T>, vararg orders: OrderSpecifier<*>): List<T> =
        factory.select(Projections.constructor(clazz))
            .from(QNotice.notice)
            .where(predicate)
            .orderBy(*orders)
            .fetch()
}